### PR TITLE
trainer: add PyTorch on AMD ROCm user guide

### PR DESCRIPTION
**Pull request by JEETDESAI25**
### Description of Changes

Adds a new user guide for running PyTorch distributed training on AMD ROCm GPUs with Kubeflow Trainer. The guide follows the same structure as the JAX on TPU guide and covers:
- PyTorch on ROCm overview (image, resources, backend)
- How to override the `torch-distributed` runtime via the Python SDK
- End-to-end MNIST DDP training example
- Reusable `ClusterTrainingRuntime` YAML for cluster admins

### Related Issues

Related: Kubeflow/trainer#2335

### Checklist

- [x] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [x] Ensure you follow best practices from our [contributing guide](https://github.com/kubeflow/website/blob/master/content/en/docs/about/contributing.md).
- [ ] (for big changes) I will post screenshots of the changes in a PR comment

### CODE REVIEW

**Review by andreyvelich — COMMENTED** [review]
Thanks for this @JEETDESAI25!
Can you sign your commits please?
/assign @astefanutti @robert-bell @abhijeet-dhumal

File: content/en/docs/components/trainer/user-guides/pytorch-rocm.md [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+job_id = client.train(
+    runtime="torch-distributed",
+    trainer=CustomTrainer(
+        func=train_mnist_rocm,
+        image="rocm/pytorch:rocm7.1.1_ubuntu24.04_py3.12_pytorch_release_2.10.0",
+        num_nodes=2,
+        resources_per_node={
+            "amd.com/gpu": 1,
+        },
+        packages_to_install=["torchvision"],
```
Comment (abhijeet-dhumal) [inline content/en/docs/components/trainer/user-guides/pytorch-rocm.md]: packages_to_install=["torchvision"] will pip-install the CUDA/CPU wheel from PyPI..
For ROCm, torchvision should come from the ROCm wheel index or be pre-baked in the image, right ?

File: content/en/docs/components/trainer/user-guides/pytorch-rocm.md [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+### Resources
+
+AMD ROCm GPUs are exposed in Kubernetes via the AMD GPU device plugin. Specify `amd.com/gpu`
+in `resources_per_node` to request AMD GPU devices:
+
+| Resource | Description |
+|----------|-------------|
+| `amd.com/gpu` | AMD GPU device managed by the AMD GPU Operator |
+
+### Tolerations
```
Comment (abhijeet-dhumal) [inline content/en/docs/components/trainer/user-guides/pytorch-rocm.md:279]: Do you think  adding node_selector not just tolerations will add value here to ensure pods land on AMD GPU nodes only incase of mixed clusters.

File: content/en/docs/components/trainer/user-guides/pytorch-rocm.md [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+                  tolerations:
+                    - key: "amd.com/gpu"
+                      operator: "Exists"
+                      effect: "NoSchedule"
+                  containers:
+                    - name: node
+                      image: rocm/pytorch:rocm7.1.1_ubuntu24.04_py3.12_pytorch_release_2.10.0
+                      resources:
+                        limits:
+                          amd.com/gpu: 1
```
Comment (abhijeet-dhumal) [inline content/en/docs/components/trainer/user-guides/pytorch-rocm.md:338]: Here both examples use amd.com/gpu: 1. Can we add.. nodes like MI300X nodes typically have 8 GPUs, and users should set this to match their hardware

File: content/en/docs/components/trainer/user-guides/pytorch-rocm.md [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+job_id = client.train(
+    runtime="torch-rocm-distributed",
+    trainer=CustomTrainer(
+        func=train_mnist_rocm,
+        packages_to_install=["torchvision"],
+    ),
+)
+```
+
+---
```
Comment (abhijeet-dhumal) [inline content/en/docs/components/trainer/user-guides/pytorch-rocm.md:353]: Maybe can add an "Environment Variables" section covering common ROCm knobs, wdyt? :
NCCL_SOCKET_IFNAME, HIP_VISIBLE_DEVICES, HSA_FORCE_FINE_GRAIN_PCIE.

**Comment by abhijeet-dhumal** [comment]
Great work @JEETDESAI25 ! 🚀 
Left a few nits, otherwise LGTM.. Thanks !

File: content/en/docs/components/trainer/user-guides/pytorch-rocm.md (hunk 1 above)
Comment (JEETDESAI25) [reply content/en/docs/components/trainer/user-guides/pytorch-rocm.md]: Yes, removed packages_to_install=["torchvision"] from both examples. The rocm/pytorch image already includes a compatible torchvision, and installing from PyPI can pull a non-ROCm wheel.

File: content/en/docs/components/trainer/user-guides/pytorch-rocm.md (hunk 2 above)
Comment (JEETDESAI25) [reply content/en/docs/components/trainer/user-guides/pytorch-rocm.md:279]: Added a ### Node Selector (optional) subsection for mixed clusters, with an example AMD GPU node label (amd.com/gpu.present: "true") for scheduling onto AMD GPU nodes.

File: content/en/docs/components/trainer/user-guides/pytorch-rocm.md (hunk 3 above)
Comment (JEETDESAI25) [reply content/en/docs/components/trainer/user-guides/pytorch-rocm.md:338]: Updated the Resources table to note that amd.com/gpu should match the number of GPUs exposed on the node hardware.

File: content/en/docs/components/trainer/user-guides/pytorch-rocm.md (hunk 4 above)
Comment (JEETDESAI25) [reply content/en/docs/components/trainer/user-guides/pytorch-rocm.md:353]: Yes sounds right, added a ### Environment Variables section documenting HIP_VISIBLE_DEVICES, NCCL_SOCKET_IFNAME, and HSA_FORCE_FINE_GRAIN_PCIE.

**Comment by JEETDESAI25** [comment]
Thanks a lot for the review, appreciate it! @abhijeet-dhumal

**Comment by abhijeet-dhumal** [comment]
/lgtm
Awesome work @JEETDESAI25, this is great addition! 🚀

**Review by andreyvelich — COMMENTED** [review]
Thanks for this @JEETDESAI25!
/lgtm
/approve
